File: backend/src/ml/predictor.py
import os
import json
import numpy as np
import joblib
from datetime import datetime, timedelta
from supabase import create_client
from dotenv import load_dotenv

from data_loader import load_daily_revenue
from features import engineer_features, to_matrix, FEATURE_KEYS
from models import RobustScaler

load_dotenv()
MODELS_DIR = os.path.join(os.path.dirname(__file__), 'savedModels')


def load_artifacts():
    """Load saved model and scaler"""
    model = joblib.load(os.path.join(MODELS_DIR, 'best_model_sklearn.pkl'))
    
    with open(os.path.join(MODELS_DIR, 'scaler.json'), 'r') as f:
        scaler_json = json.load(f)
    scaler = RobustScaler.from_json(scaler_json)
    
    with open(os.path.join(MODELS_DIR, 'feature_names.json'), 'r') as f:
        feature_keys = json.load(f)
    
    return model, scaler, feature_keys


def build_future_rows(daily, n_future=14):
    """Create placeholder rows for future dates using recent trends"""
    if len(daily) == 0:
        return []
    
    # Use last 7 days with revenue > 0 for patterns
    recent = [r for r in daily[-14:] if r['revenue'] > 0]
    if len(recent) == 0:
        recent = daily[-7:] if len(daily) >= 7 else daily
    
    avg_sessions = sum(r['sessions'] for r in recent) / len(recent)
    avg_viewers = sum(r['viewers'] for r in recent) / len(recent)
    avg_likes = sum(r['likes'] for r in recent) / len(recent)
    
    # Calculate trend from last 3 revenue days
    revenue_days = [r for r in daily if r['revenue'] > 0]
    if len(revenue_days) >= 3:
        last_three = revenue_days[-3:]
        revenues = [r['revenue'] for r in last_three]
        trend = (revenues[-1] - revenues[0]) / 2
    else:
        trend = 0
    
    max_period = max((r.get('period_id') or 0 for r in daily), default=0)
    last_date = datetime.strptime(daily[-1]['date'], '%Y-%m-%d')
    
    future_rows = []
    last_revenue = revenue_days[-1]['revenue'] if revenue_days else 50000000
    
    for i in range(n_future):
        d = last_date + timedelta(days=i + 1)
        
        # Apply trend decay (assuming revenue continues trend but slows)
        predicted_revenue = max(0, last_revenue + (trend * (i + 1) * 0.7))
        
        future_rows.append({
            'date': d.strftime('%Y-%m-%d'),
            'revenue': predicted_revenue,  # This is what we're predicting
            'sessions': max(1, round(avg_sessions * (1 - i * 0.05))),
            'viewers': max(1, round(avg_viewers * (1 - i * 0.05))),
            'likes': max(1, round(avg_likes * (1 - i * 0.05))),
            'period_id': max_period + i + 1,
        })
        
        last_revenue = predicted_revenue
    
    return future_rows


def predict_and_save(brand_id=None, n_future=14):
    supabase = create_client(
        os.getenv('SUPABASE_URL'),
        os.getenv('SUPABASE_SERVICE_KEY')
    )
    
    model, scaler, feature_keys = load_artifacts()
    historical = load_daily_revenue(brand_id)
    
    if len(historical) == 0:
        return {'error': 'No historical data found'}
    
    print(f"Historical days: {len(historical)}")
    print(f"Days with revenue: {len([r for r in historical if r['revenue'] > 0])}")
    
    future_rows = build_future_rows(historical, n_future)
    print(f"Future days to predict: {len(future_rows)}")
    
    # Show sample future predictions
    for fr in future_rows[:3]:
        print(f"  Future period {fr['period_id']}: predicted base = Rp {fr['revenue']:,.0f}")
    
    all_rows = historical + future_rows
    
    featured = engineer_features(all_rows)
    X = to_matrix(featured, feature_keys)
    X_scaled = scaler.transform(X)
    
    raw_preds = model.predict(np.array(X_scaled)).tolist()
    
    # Clear existing predictions
    supabase.table('revenue_predictions').delete().neq('id', 0).execute()
    
    records = []
    for i, row in enumerate(all_rows):
        is_future = i >= len(historical)
        
        # For future rows, use the model prediction
        # For historical, use actual revenue
        if is_future:
            predicted_value = max(0, round(raw_preds[i]))
        else:
            predicted_value = row['revenue']
        
        records.append({
            'period_id': row.get('period_id') or 0,
            'period_name': f"Period {row.get('period_id') or 0}",
            'date': row['date'],
            'actual': None if is_future else row['revenue'],
            'predicted': predicted_value,
            'is_future': is_future,
            'model_r2': 0.84,
            'model_mae': 5121462,
            'model_slope': None,
        })
        
        if is_future:
            print(f"  Period {row.get('period_id')}: ML predicted = Rp {predicted_value:,.0f}")
    
    for i in range(0, len(records), 500):
        batch = records[i:i+500]
        supabase.table('revenue_predictions').insert(batch).execute()
    
    print(f"\nSaved {len(records)} predictions to Supabase")
    print(f"   Historical: {len(historical)} days")
    print(f"   Future predictions: {len(future_rows)} days")
    
    return {
        'saved': len(records),
        'historical': len(historical),
        'future': n_future,
    }


if __name__ == '__main__':
    result = predict_and_save()
    print(json.dumps(result, indent=2))

File: backend/src/ml/predictor.js
// predictor.js - ES Module version
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';
import { loadDailyRevenue } from './data_loader.js';
import { engineerFeatures, toMatrix, FEATURE_KEYS } from './features.js';
import { RobustScaler, RFRegressor, RidgeRegression } from './models.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.join(__dirname, '../../.env') });

const MODELS_DIR = path.join(__dirname, 'savedModels');

function loadArtifacts() {
  const modelJson = JSON.parse(fs.readFileSync(path.join(MODELS_DIR, 'best_model.json'), 'utf8'));
  const scalerJson = JSON.parse(fs.readFileSync(path.join(MODELS_DIR, 'scaler.json'), 'utf8'));
  const featureKeys = JSON.parse(fs.readFileSync(path.join(MODELS_DIR, 'feature_names.json'), 'utf8'));
  const modelType = JSON.parse(fs.readFileSync(path.join(MODELS_DIR, 'model_type.json'), 'utf8'));

  const scaler = RobustScaler.fromJSON(scalerJson);

  let model;
  if (modelType.type === 'RFRegressor') {
    model = RFRegressor.fromJSON(modelJson);
  } else {
    model = RidgeRegression.fromJSON(modelJson);
  }

  return { model, scaler, featureKeys };
}

function buildFutureRows(daily, nFuture = 14) {
  if (daily.length === 0) return [];

  const recent = daily.slice(-14).filter(r => r.revenue > 0);
  const base = recent.length > 0 ? recent : (daily.slice(-7).length > 0 ? daily.slice(-7) : daily);

  const avgSessions = base.reduce((s, r) => s + r.sessions, 0) / base.length;
  const avgViewers = base.reduce((s, r) => s + r.viewers, 0) / base.length;
  const avgLikes = base.reduce((s, r) => s + r.likes, 0) / base.length;

  const revenueDays = daily.filter(r => r.revenue > 0);
  let trend = 0;
  if (revenueDays.length >= 3) {
    const lastThree = revenueDays.slice(-3);
    const revenues = lastThree.map(r => r.revenue);
    trend = (revenues[revenues.length - 1] - revenues[0]) / 2;
  }

  const maxPeriod = Math.max(...daily.map(r => r.period_id || 0), 0);
  const lastDate = new Date(daily[daily.length - 1].date);
  let lastRevenue = revenueDays.length > 0 ? revenueDays[revenueDays.length - 1].revenue : 50000000;

  const futureRows = [];
  for (let i = 0; i < nFuture; i++) {
    const d = new Date(lastDate);
    d.setDate(d.getDate() + i + 1);
    const predictedRevenue = Math.max(0, lastRevenue + trend * (i + 1) * 0.7);

    futureRows.push({
      date: d.toISOString().split('T')[0],
      revenue: predictedRevenue,
      sessions: Math.max(1, Math.round(avgSessions * (1 - i * 0.05))),
      viewers: Math.max(1, Math.round(avgViewers * (1 - i * 0.05))),
      likes: Math.max(1, Math.round(avgLikes * (1 - i * 0.05))),
      period_id: maxPeriod + i + 1,
    });

    lastRevenue = predictedRevenue;
  }

  return futureRows;
}

export async function predictAndSave(brandId = null, nFuture = 14) {
  const supabase = createClient(
    process.env.SUPABASE_URL,
    process.env.SUPABASE_SERVICE_KEY
  );

  const { model, scaler, featureKeys } = loadArtifacts();
  const historical = await loadDailyRevenue(brandId);

  if (historical.length === 0) {
    return { error: 'No historical data found' };
  }

  console.log(`Historical days: ${historical.length}`);
  console.log(`Days with revenue: ${historical.filter(r => r.revenue > 0).length}`);

  const futureRows = buildFutureRows(historical, nFuture);
  console.log(`Future days to predict: ${futureRows.length}`);

  for (const fr of futureRows.slice(0, 3)) {
    console.log(`  Future period ${fr.period_id}: predicted base = Rp ${fr.revenue.toLocaleString()}`);
  }

  const allRows = [...historical, ...futureRows];
  const featured = engineerFeatures(allRows);
  const X = toMatrix(featured, featureKeys);
  const XScaled = scaler.transform(X);

  const rawPreds = model.predict(XScaled);

  await supabase.table('revenue_predictions').delete().neq('id', 0);

  const records = allRows.map((row, i) => {
    const isFuture = i >= historical.length;
    const predictedValue = isFuture ? Math.max(0, Math.round(rawPreds[i])) : row.revenue;

    if (isFuture) {
      console.log(`  Period ${row.period_id}: ML predicted = Rp ${predictedValue.toLocaleString()}`);
    }

    return {
      period_id: row.period_id || 0,
      period_name: `Period ${row.period_id || 0}`,
      date: row.date,
      actual: isFuture ? null : row.revenue,
      predicted: predictedValue,
      is_future: isFuture,
      model_r2: 0.84,
      model_mae: 5121462,
      model_slope: null,
    };
  });

  for (let i = 0; i < records.length; i += 500) {
    const batch = records.slice(i, i + 500);
    await supabase.table('revenue_predictions').insert(batch);
  }

  console.log(`\nSaved ${records.length} predictions to Supabase`);
  console.log(`   Historical: ${historical.length} days`);
  console.log(`   Future predictions: ${futureRows.length} days`);

  return {
    saved: records.length,
    historical: historical.length,
    future: nFuture,
  };
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  predictAndSave().then(result => console.log(JSON.stringify(result, null, 2)));
}